File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    def helper(self,money):
        l=[0]*len(money)
        l[0]=money[0]
        for i in range(1,len(money)):
            pick=money[i]
            if i-2>=0:
                pick+=l[i-2]
            notpick=l[i-1]
            l[i]=max(pick,notpick)
        return l[-1]

    def rob(self, nums: List[int]) -> int:
        if len(nums)==1:
            return nums[0]
        ans1=self.helper(nums[:-1])
        ans2=self.helper(nums[1:])
        return max(ans1,ans2)
        

File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {
    public int helper(int start,int end,int[] nums){
        if(end-start+1==1){
            return nums[start];
        }
        int first=nums[start];
        int second=Math.max(nums[start+1],first);
        for(int i=start+2;i<=end;i++){
            int temp=Math.max(second,nums[i]+first);
            first=second;
            second=temp;
        } 
        return second;
    }
    public int rob(int[] nums) {
        if(nums.length==1){
            return nums[0]; 
        }
        return Math.max(helper(0,nums.length-2,nums),helper(1,nums.length-1,nums));
    }
}